controllers/teleport_controller/teleport_controller.cpp

```diff
@@ -0,0 +1,138 @@
+// File:          teleport_controller.cpp
+// Date:          23/03/2021
+// Description:   Method to teleport robot around the playing field for the purposes of
+//                    gathering training data
+// Author:        Lachlan Court
+// Modifications:
+
+
+#include <RobotisOp2GaitManager.hpp>
+#include <RobotisOp2MotionManager.hpp>
+#include <webots/Supervisor.hpp>
+#include <iostream>
+#include <math.h>
+#include <time.h> 
+
+// All the webots classes are defined in the "webots" namespace
+using namespace webots;
+
+/*
+// Rotations matrix for 90 degree rotations
+static const int rotationsSize = 4;
+static const double rotations[rotationsSize][4] = {{1, 0, 0, 1.57081},
+                                                  {0.577353, 0.577345, 0.577353, 2.0944},
+                                                  {1.69526e-09, 0.707103, 0.707111, 3.14159},
+                                                  {-0.577352, 0.577347, 0.577352, -2.09441}};
+
+*/
+
+// Rotations matrix for 45 degree rotations
+static const int rotationsSize = 8;                                     
+static const double rotations[rotationsSize][4] = {{1, 0, 0, 1.57081},
+                                                  {0.867737, 0.343283, 0.359429, 1.75753},
+                                                  {0.577353, 0.577345, 0.577353, 2.0944},
+                                                  {0.286947, 0.661632, 0.692752, 2.60662},
+                                                  {1.69526e-09, 0.707103, 0.707111, 3.14159},
+                                                  {-0.286948, 0.661631, 0.692752, -2.60661},
+                                                  {-0.577352, 0.577347, 0.577352, -2.09441},
+                                                  {0.867739, -0.343277, -0.359429, 1.75753}};                               
+
+
+/* This is the main program of the supervisor. It creates an instance of the Supervisor
+ *     instance, launches its function(s) and destroys it at the end of the execution.
+ */
+int main() {
+    std::cout << "a" << std::endl;
+    // create the Supervisor instance and assign it to a robot
+    Supervisor* supervisor = new Supervisor();
+    Node* target           = supervisor->getFromDef("RED_1");
+    std::cout << "b" << std::endl;
+    // Get the time step of the current world.
+    int timeStep = (int) supervisor->getBasicTimeStep();
+
+    std::cout << "c" << std::endl;
+
+    // Add the motion manager to make the robot stand up and look around
+    managers::RobotisOp2MotionManager* mMotionManager = new managers::RobotisOp2MotionManager(supervisor);
+    mMotionManager->playPage(54);
+
+    // Generate random seed
+    srand(time(NULL));
+
+    // Main loop:
+    while (supervisor->step(timeStep) != -1) {
```

Comment:
This loop condition confuses me. It should probably be reworked, or at least commented explaining why this exotic loop condition is necessary

Reply:
https://cyberbotics.com/doc/guide/controller-programming
See part about `wb_robot_step`. 
The main loop is a specific thing in controllers in which each time step the controller code executes everything needed in that time step. Otherwise, without a loop, it'll just execute and stop, and the robot won't do anything. 
`step(timeStep)` is probably similar to that function in the docs, where the function takes `timeStep` seconds to execute and so gives a delay between each execution of the while loop. Why `-1` is another question, I guess that's what it returns when all is ok? Seems to be standard in webots.

Reply:
Yep this was just the format of the sample code, it runs everything in that loop repeatedly until the simulation is terminated